### DISCUSSION TITLE
Add keyed options to dependency injection

### DISCUSSION THREAD

**Discussion by Dovencio** [dotnet/runtime#101097] [Ideas]
With the current implementation of DI there is no easy way of configuring a keyed service. I see two ways of implementing keyed options:
Add `ConfigureKeyed<TOption>(object? Key, // ...)`:
```
public class Service
{
    private readonly IOptions<ServiceOptions> options;
    // ...
    public Service(IOptions<ServiceOptions> options, // ...)
    {
        this.options = options;
        // ...
    }
}

public class ServiceOptions
{
    public string Property { get; set; }
    // ...
}

provider.Services.AddKeyedScoped<Service>("ServiceKeyA");
provider.Services.AddKeyedScoped<Service>("ServiceKeyB");
provider.Services.ConfigureKeyed<Service>("ServiceKeyA", o => {
    o.Property = "foo";
});
// ...
```
Or by passing the key into the constructor and use named configurations:
```
public class Service
{
    private readonly ServiceOption options;
    // ...
    public Service([ServiceKey] object? key, IOptionSnapshot<ServiceOptions> accessor)
    {
        string? serviceKey = key as string;
        if (serviceKey is null) throw Exception($"Service key {key} is null or not a string!");
        options = accessor.Get(serviceKey);
    }
}

public class ServiceOptions
{
    public string Property { get; set; }
    // ...
}

provider.Services.AddKeyedScoped<Service>("ServiceKeyA");
provider.Services.AddKeyedScoped<Service>("ServiceKeyB");
provider.Services.Configure<Service>("ServiceKeyA", o => {
    o.Property = "foo";
});
// ...
```
Personally I would prefer `ConfigureKeyed` function as it could provide multiple ways to configure as well as adding named configurations like `ConfigureKeyed<Service>("ServiceA", "MyName", o => { // ... });` rather than use `Configure<Service>("ServiceA.MyName", o => { // ... });`. 

Additionally using both `ConfigureKeyed` and `Configure` could make default options.
```
partial class Service
{
    string GetProperty() => options.Property;
}
// ...
provider.Services.AddScoped<Service>();
provider.Services.AddKeyedScoped<Service>("ServiceKeyA");
provider.Services.AddKeyedScoped<Service>("ServiceKeyB");
provider.Services.Configure<Service>(o => {
    o.Property = "Hello from default Service Config!";
});
provider.Services.ConfigureKeyed<Service>("ServiceKeyA", o => {
    o.Property = "Hello from ServiceKeyA Config!";
});
// ...

Service serveA = app.Services.GetKeyedService<Service>("ServiceKeyA");
Service serveB = app.Services.GetKeyedService<Service>("ServiceKeyB");
Service serveC = app.Services.GetService<Service>();

Console.WriteLine(serveA.GetProperty()); // "Hello from ServiceKeyA Config!"
Console.WriteLine(serveB.GetProperty()); // "Hello from default Service Config!"
Console.WriteLine(serveC.GetProperty()); // "Hello from default Service Config!"
```

**Comment by tacosontitan** [comment]
What prevents you from using abstractions as keys instead?

```cs
public interface ISampleService;
public interface IAlphaSampleService : ISampleService;
public interface IBravoSampleService : ISampleService;

...
services.AddScoped<IAlphaSampleService, AlphaSampleService>()
    .AddScoped<IBravoSampleService, BravoSampleService>()
    .AddScoped<ISampleService, DefaultSampleService>();
    
...
public sealed class Foo(IAlphaSampleService sampleService);
```

IMHO I prefer to have compile time enforcement of my services. If you're truly set on using runtime keys instead, then you could always create a factory for resolving things and an extension method on `IServiceProvider` that exposes it for consumption:

```cs
public interface IKeyedServiceFactory
{
    // just an example
    TService GetService<TService>(object key);
}

...
public static TService GetKeyedService<TService>(this IServiceProvider source, object key)
{
    var factory = source.GetRequiredService<IKeyedServiceFactory>();
    return factory.GetService<TService>(key);
}

...
var service = serviceProvider.GetKeyedService<SampleService>("alpha");
```

Though, again, I don't recommend this and will leave further exploration to the reader.

**Reply by Dovencio** [reply]
I'm having a frustration with `IEmailSender`'s unable to pass in send from `SendAsync`. For example:
appsettings.json:
```
{
    "EmailClients": {
        "no-reply": {
            "email": "no-reply@example.com",
            "name": "no-reply"
        },
        "support": {
            "email": "support@example.com",
            "name": "support"
        }
    }
}
```
Program.cs:
```
// ...
foreach (IConfigurationSection child in builder.Configuration.GetSection("EmailClients").GetChildren())
{
    if (child.Key == "default")
    {
        builder.Services.AddScoped<IEmailSender, EmailSender>();
        builder.Services.Configure<EmailOtions>(defaultConfig);
    }
    else
    {
        builder.Services.AddKeyedScoped<IEmailSender, EmailSender>(child.Key);
        builder.Services.ConfigureKeyed<EmailOptions>(child.Key, child);
    }
}
// ...
```
controller.cs:
```
public class EmailSenderController(IServiceProvider provider) : PageController
{
    public async Task SendOnPost(string from, string to, string message)
    {
        IEmailSender sender = provider.GetKeyedService<IEmailSender>(from);
        sender ?= provider.GetRequiredService<IEmailSender>();
        await sender.SendEmailAsync(to, "Your car's extended warranty", message);
   }
}
```
This way if I want to add `abuse@example.com` I don't have to recompile and publish the whole thing, I just need to add the new entry into appsettings.json.

**Reply by tacosontitan** [reply]
This screams X/Y problem to me. Personally, I'd recommend using something like:

```cs
public interface IEmail;
public sealed class SupportEmail(IOptionsMonitor<Sample> options)
{
    ... get support email details from appsettings here
}

...
var emailProvider = provider.GetRequiredSerivce<IEmailProvider>();
var email = new SupportEmail(to, subject, message);
await emailProvider.Send(email, cancellationToken);
```

**Reply by Dovencio** [reply]
This is where you've lost me because I've already solved that specific issue and my solution doesn't look like yours. I'm not here to look for a solution, I'm here to offer an idea that I believe would help streamline what I did to solve the issue. There can't be an XY problem if Y has already been solved. I should rephrase to be clear, if the functions `AddKeyed{Scope}<T>(object? key)` exists, why doesn't/shouldn't the function `ConfigureKeyed<T>(object? key, ...)` exist?

**Reply by tacosontitan** [reply]
Ah, I now see that this is an issue of me missing a critical component of your question, *configuration* as opposed to *registration*. My apologies!

**Comment by SerratedSharp** [comment]
I really miss keyed services.  If you have an interface with multiple concrete implementations, and you want DB or config file driven selection of the implementation, they are really great.  I can implement handlers for a dozen different payment vendors, and in a multi tenant system allow my clients to configure their payment vendor, and have the appropriate implementation light up.  Done properly all the other code on each side of the pipeline leverages the interface, and the configuration drives the selection of the concrete implementation.  They are really great when used properly.  When you show another dev how easy it is with a bit of declarative config to switch implementations, and eliminate repetitive blocks of switch cases and redundant concrete classes, they begin to really appreciate this approach.

Usually counter examples that don't understand how to implement this properly will hard code a string during the service resolution.  That's the wrong way to leverage this approach.  You would pass a string, but it wouldn't be hardcoded, it would be determined at runtime based on some dynamic or data driven process.

HttpClient was another great use of them.  You generally want a single HttpClient instead per integration. But you don't want one HttpClient for the entire app, because there's alot of things like timeout property that cannot be changed on HttpClient after the first request.  If you have multiple integrations, then you want one HttpClient per integration, so the BaseUri, Timeout, etc. can be specific to the remote integration.  Keyed service registration solved this problem cleanly.  Creating a bunch of class declarations for what differs only by some configuration is unnecessary bloat, and results in the consuming code being overly coupled to what would have otherwise just been an HttpClient.  They all share the same interface, there's no need for different types.  For this use case, having attributes on constructor parameters that name the desired implementation made more sense, because you usually wanted the specifically configured HttpClient singleton for a particular section of code.  The big difference here is there aren't multiple implementations, but instead there's a single implementation with multiple singleton instances.  I.e. if I have a integration with Google API and one with OpenAI, then I need a singleton HttpClient for Google API, and another singleton HttpClient for OpenAI.  

I thought it was real interesting when we got the extension methods for .AddHttpClient that take a name, which is basically exposing them as keyed service outside of DI. I felt vindicated. It would have made more sense to just fix the DI though IMO.

**Reply by rjgotten** [reply]
Keyed services pollute a DI container though. And nothing prevents any other developer on a team from calling out to them in unintended ways.

You're better off having an interface, say `IFoo`, with a `CompositeFoo` implementation registered in the service collection, where `CompositeFoo` takes an `IOptionsSnapshot<FooOptions>` that configures the available concrete types. `CompositeFoo` could create efficient reusable factory delegates for these concrete types via `ActivatorUtilities.CreateFactory`, which would pull their constructor parameters from the service provider but crucially would _not_ require the concrete implementations themselves to be reachable through it. And then it could itself simply be responsible for calling the correct factory delegate to get an instance of whatever concrete type it deems is required for the situation, and forward the handling of the `IFoo` interface contract to that implementation.

**Reply by SerratedSharp** [reply]
I would discourage others from rolling their own factory when there's plenty of implementations that are well refined.  When you go down the road of building your own factory there will be gaps, shortcomings and missteps in the design that more mature implementations have all addressed.  I always hate inheriting code bases where someone has reinvented the wheel on cross cutting concerns.  There's several DI container implementations that are superior, mature, and support keyed containers.  There's a reason they have this feature.  I think its hilarious that they had to work around the gap in .NET runtime's implementation with the HttpClient keyed container.  You can call it what you want, but that's essentially what it is.

"And nothing prevents any other developer on a team from calling out to them in unintended ways."
Already true of any other non-keyed service.  You have the same APIs available for on-demand instantiation versus constructor/attribute based injection.  Nothing changes there.

**Reply by rjgotten** [reply]
>Already true of any other non-keyed service.

That is incorrect.
In the factory solution you register a concrete composite service under the service interface, which knows how to delegate to particular different implementations and under which conditions. In the other, you register all the implementations under the service interface directly in the DI container as keyed services, you make the callers responsible for knowing the selection logic, and you have nothing precluding a developer from just picking one of them arbitrarily by its known key - bypassing all intended constraints, business logic, system logic, or otherwise.

**Comment by SerratedSharp** [comment]
You're building straw men by describing the worst possible way to leverage
a keyed containers.

When correctly leveraged the keyed access is usually configuration or data
driven.  This is exactly how multi tenants systems implement support for
multiple payment providers for example.  They can all share the same
interface for multiple implementations.  No one knows how to leverage
interfaces properly anymore because they are locked into this fallacy that
there's only one and exactly one concrete implementation of every interface.

**Reply by rjgotten** [reply]
> This is exactly how multi tenants systems implement support for multiple payment providers for example.  

Multi-tenant solutions will more likely be using a DI framework capable of supporting child containers. That enables them to load different implementations for different tenants, as well as give those implementations a multiton lifetime scope, basically: a singleton per tenant, with a lot less hassle than managing keyed services.
